include/ghex/transport_layer/mpi/communicator.hpp

```diff
@@ -28,162 +30,226 @@ namespace gridtools {
-                      * @param source source rank
-                      * @param tag message tag
-                      * @return completion handle */
                     template<typename Message>
-                    [[nodiscard]] future<void> recv(Message& msg, rank_type source, tag_type tag) const
-                    {
+                    [[nodiscard]] future<void> recv(Message& msg, rank_type src, tag_type tag) const {
                         request req;
-                        GHEX_CHECK_MPI_RESULT(
-                                MPI_Irecv(reinterpret_cast<void*>(msg.data()),sizeof(typename Message::value_type)*msg.size(), 
-                                          MPI_BYTE, source, tag, *this, &req.get()));
+                        GHEX_CHECK_MPI_RESULT(MPI_Irecv(reinterpret_cast<void*>(msg.data()),
+                                                        sizeof(typename Message::value_type) * msg.size(), MPI_BYTE,
+                                                        src, tag, m_shared_state->m_comm, &req.get()));
                         return req;
                     }
 
-                    /** @brief non-blocking receive which allocates the container within this function and returns it
-                      * in the future 
-                      * @tparam Message a container type
-                      * @tparam Args additional argument types for construction of Message
-                      * @param n number of elements to be received
-                      * @param source source rank
-                      * @param tag message tag
-                      * @param args additional arguments to be passed to new container of type Message at construction 
-                      * @return completion handle with message as payload */
-                    template<typename Message, typename... Args>
-                    [[nodiscard]] future<Message> recv(int n, rank_type source, tag_type tag, Args&& ...args) const
+                    template <typename MsgType, typename Neighs>
+                    std::vector<future<void>> send_multi(MsgType& msg, Neighs const &neighs, tag_type tag) const {
+                        std::vector<future<void>> res;
+                        res.reserve(neighs.size());
+                        for (auto id : neighs)
+                            res.push_back( send(msg, id, tag) );
+                        return res;
+                    }
+
+                    unsigned progress() { return m_state->progress(); }
+
+                    void barrier() {
+                        if (auto token_ptr = m_state->m_token_ptr) {
+                            auto& tp = *(m_shared_state->m_thread_primitives);
+                            auto& token = *token_ptr;
+                            tp.barrier(token);
+                            tp.single(token, [this]() { MPI_Barrier(m_shared_state->m_comm); } );
+                            tp.barrier(token);
```

Comment:
Why do you need a second thread barrier?

Reply:
I am not sure if we need it here. We certainly need it for the ucx communicator. Maybe @angainor can comment as well.

Reply:
In UCX we need both to make sure that only a single thread progresses the communication. Here we probably don't, but it's a question of which one do you want to keep: the first one, or the second one. I'd probably keep the second one only: let the first thread start the global barrier immediately, without having to wait for all the others. Could be more efficient, but then again we are not targeting efficiency here.